File: 2-EST/Bag.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#-----------------------------------------------------------------------
# Bag.py
# An implementation of the Bag supported on a python list
#
# Author: Jorge Londoño
#-----------------------------------------------------------------------



class Bag:
    """Implementacion de un Bag por medio de una lista
    """
    
    # Constructor
    def __init__(self):
        """Constructor: Inicializa la variable de instancia como la lista vacia
        """
        self._datos = []

    
    def add(self, item):
        """Agrega un item a la bolsa
        """
        self._datos.append(item)


    def size(self):
        """Devuelve el numero de elementos en el Bag
        """
        return len(self._datos)


    def isEmpty(self):
        """Comprueba si el Bag se encuentra vacio
        """
        return self.size()==0

        
    # Iterador basado en el iterador de la lista 
    def __iter__(self):
        """Devuelve un iterador del Bag reutilizando el iterador del list
        """
        return iter(self._datos)



# Aplicacion cliente ejemplo
if __name__ == '__main__':

    # Crear una bolsa y agregar algunos items    
    bolsa = Bag();
    print("isEmpty = {0}".format(bolsa.isEmpty()))
    bolsa.add('Hola')
    bolsa.add('Mundo')
    print("size = {0}".format(bolsa.size()))
    print("isEmpty = {0}".format(bolsa.isEmpty()))

    # Iterar sobre los items de la bolsa
    for x in bolsa:
        print(x)

File: 2-EST/Bag.ts
/**
 * Bag: Implementación de la bolsa en TypeScript
 * 
 * @author: Jorge Londoño
 */

import assert from "assert";


class Bag<T> implements Iterable<T> {

    private items: Array<T> = [];
    private n: number = 0;

    add(item: T): void { 
        this.items.push(item);
        this.n++;
    }

    isEmpty(): boolean {
        return this.n==0;
    }

    size(): number {
        return this.n;
    }

    [Symbol.iterator]() {
        let pos = 0;
        const items = this.items;

        return {
            next(): IteratorResult<T> {
                if (pos<items.length) {
                    return { value: items[pos++], done: false };
                }
                return { value: null, done: true };
            }
        }
    }

    static factory<T>(): Bag<T> {
        return new Bag<T>();
    }

}


// Pruebas unitarias de Bag
const bolsaStr: Bag<string> = new Bag();
assert.ok(bolsaStr.isEmpty(), "La bolsa debe estar vacia al comienzo");
assert.strictEqual(bolsaStr.size(), 0, "Tamaño inicial debe ser cero");
bolsaStr.add("Hola");
bolsaStr.add("Mundo");
assert.equal(bolsaStr.size(), 2, "Tamaño incorrecto");


// Cliente ejemplo de Bag
const bolsaNum: Bag<number> = new Bag();
bolsaNum.add(1);
bolsaNum.add(2);
bolsaNum.add(3);

for(let x of bolsaNum) {
    console.log(x);
}
